keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1165 @@
+type JobStatus struct {
+  ...
+
+  // FailedIndexes holds the failed indexes when
+  // `.spec.runPolicy.backoffLimitPerIndex=true`.
+  // The indexes are represented in the text format analogous as for the
+  // `completedIndexes` field, ie. they are kept as decimal integers
+  // separated by commas. The numbers are listed in increasing order. Three or
+  // more consecutive numbers are compressed and represented by the first and
+  // last element of the series, separated by a hyphen.
+  // For example, if the failed indexes are 1, 3, 4, 5 and 7, they are
+  // represented as "1,3-5,7".
+  // +optional
+  FailedIndexes *string
+}
+```
+
+### Tracking the number of failures per index
+
+In order to determine if the backoff limit per index is exceeded we keep
+track of the number of failures per index. For this purpose we use the Pod
+annotation, `batch.kubernetes.io/job-index-failure-count`, which holds the value
+of the number of pod failures for a given index. It is set to `0` for the first
+pod created for a given index.
+
+When Job controller sees a failed pod corresponding to a given index, and the
+value of the annotation `batch.kubernetes.io/job-index-failure-count` is greater
+or equal to the configured backoff limit per index then the index is marked
+as failed and added to `.status.failedIndexes`.
+
+When Job controller creates replacement pods for failed pods for a given
+index it checks if the index isn't finished yet (it is not in
+`.status.failedIndexes` nor `.status.completedIndexes`).
+Then, if `x` is the highest `batch.kubernetes.io/job-index-failure-count`
+for the index, the newly created pod will have the annotation set to `x+1`.
+
+In order to keep track of the number of failures per index, the Job controller
+removes finalizers of a failed pod for a given index, only once the replacement
+pod (with incremented value of `batch.kubernetes.io/job-index-failure-count`) is
+created, or the index is marked as failed in `.status.failedIndexes`.
+
+### Failed indexes format
+
+The format of the `.status.failedIndexes` field is analogous to the one used for
+successful indexes, represented by the `completedIndexes` field, which is a
+text format grouping consecutive integers into ranges. In a special case, when
+the indexes are non-consecutive they are represented by comma-separated numbers.
+In the worst-case scenario this is a string of comma-separated even values from
+`0` to `.spec.completions-1`. In order to constrain the size of the field we
+constrain the number of completions to `10^5`. This is the same constraint as
+put on the parallelism for the Indexed Job to make sure that the size of
+`completedIndexes` is limited. With this limit the length of the string is
+limited by `295k` characters.
+
+### Job completion
+
+When backoff limit per index is used, then we execute indexes until all of them
+are completed (either failed or succeeded).
+
+Then, the Job is marked as completed (the `Complete` Job condition type) when
```

Comment:
I have added the note in alternatives about the semantics. Indeed marking Job as failed if any index failed might be more natural to users, for example in the context of integration tests. I may swap the current strategy with the alternative actually. @jensentanlo what is your view?

Reply:
Is there something that the user should be watching for to conclude that the indexed job has finished running? 

Would the failed condition be applied as soon as a single index fails (or a certain number of indices, if the API is extended in that way) , or would it be applied after all indexes have finished running?

Reply:
2 is my pick

Reply:
Updated.

Reply:
How do y'all feel about introducing a new conditions which would differ from complete and failed to better express the state? Just my 2 cents

Reply:
I assume it is useful to add the Failed or Complete Job conditions so that the Job object can be interpreted with an older client code. 

Maybe we could have new conditions added along with `Failed` or `Complete`, but it seems that the reason field is enough.

Reply:
I agree that we should add one of the two conditions for backwards compatibility. And if we have to do that, adding a third condition might be overkill.

Reply:
Resolving the discussion.